File: Phoenix/lib/biab_starter_web/controllers/store_controller.ex
defmodule BiabStarterWeb.StoreController do
  use BiabStarterWeb, :controller

  alias BiabStarter.Biab

  def index(conn, params) do
    grid =
      Biab.product_grid(
        search: Map.get(params, "search"),
        category_id: Map.get(params, "categoryId"),
        sort: Map.get(params, "sort")
      )

    render(conn, :index,
      products: Map.get(grid, "items", []),
      category_counts: Map.get(grid, "categoryCounts", []),
      search: Map.get(params, "search", "")
    )
  end

  def show(conn, %{"id" => id}) do
    case Biab.product(id) do
      nil ->
        conn |> put_status(:not_found) |> text("Product not found")

      product ->
        render(conn, :show,
          product: product,
          related: Biab.related_products(id),
          reviews: Biab.product_reviews(id)
        )
    end
  end
end

File: Vapor-Server/Sources/App/Controllers/StoreController.swift
import Vapor

enum StoreController {
    static func index(req: Request) async throws -> View {
        let search = try? req.query.get(String.self, at: "search")
        let sort = try? req.query.get(String.self, at: "sort")

        guard let bd = req.bd else {
            return try await req.view.render("store", StoreContext(
                configured: false, products: [], search: search ?? ""
            ))
        }

        let key = "storefront:grid:\(search ?? "")|\(sort ?? "")"
        let products = await req.bdCache.fetch(key, tags: ["bd:storefront"], fallback: [Product]()) {
            try await bd.productGrid(search: search, sort: sort).products
        }

        return try await req.view.render("store", StoreContext(
            configured: req.application.bdConfigured,
            products: products.map(ProductCard.init),
            search: search ?? ""
        ))
    }

    static func show(req: Request) async throws -> View {
        guard let bd = req.bd, let id = req.parameters.get("id") else { throw Abort(.notFound) }

        let product = await req.bdCache.fetch(
            "storefront:product:\(id)", tags: ["bd:storefront"], fallback: Optional<Product>.none
        ) { try await bd.product(id) }

        guard let product else { throw Abort(.notFound) }

        let reviews = await req.bdCache.fetch(
            "storefront:reviews:\(id)", tags: ["bd:storefront"], fallback: [Review]()
        ) { try await bd.productReviews(id).reviews }

        return try await req.view.render("product", ProductContext(
            configured: req.application.bdConfigured,
            id: product.id,
            name: product.name,
            description: product.description ?? "",
            price: Money.cents(product.cheapestPriceCents),
            reviews: reviews
        ))
    }
}
